src/Npgsql/TypeHandlers/UnmappedCompositeHandler.cs

```diff
@@ -192,10 +192,13 @@ protected internal override Task WriteObjectWithLength(object value, NpgsqlWrite
                 ? WriteWithLengthInternal<DBNull>(null, buf, lengthCache, parameter, async)
                 : WriteWithLength(value, buf, lengthCache, parameter, async);
 
-        protected override Task WriteWithLength<T2>(T2 value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        protected override async Task WriteWithLength<T2>(T2 value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
         {
+            if (buf.WriteSpaceLeft < 4)
+                await buf.Flush(async);
+
             buf.WriteInt32(ValidateAndGetLength(value, ref lengthCache, parameter));
-            return Write(value, buf, lengthCache, parameter, async);
+            await Write(value, buf, lengthCache, parameter, async);
```

Comment:
Should this branch on the `async` parameter?

Reply:
Will do.

Reply:
Just remember why I wrote it this way. An exception thrown by `ValidateAndGetLength` won't be wrapped by `Task`, but it must. There are few other places where it should be fixed.